File: fixtures/scale/scala/medium/src/calc/parser/Precedence.scala
package calc.parser

import calc.lexer.TokenType

object Precedence {
  val Lowest = 0

  def of(tokenType: TokenType): Int = tokenType match {
    case TokenType.Plus | TokenType.Minus                    => 10
    case TokenType.Star | TokenType.Slash | TokenType.Percent => 20
    case TokenType.Caret                                     => 30
    case _                                                   => Lowest
  }

  def rightAssociative(tokenType: TokenType): Boolean =
    tokenType == TokenType.Caret

  /** The minimum precedence for the right-hand side of the given operator. */
  def nextFor(tokenType: TokenType): Int =
    if (rightAssociative(tokenType)) of(tokenType) - 1 else of(tokenType)
}

File: fixtures/scale/java/medium/src/calc/parser/Precedence.java
package calc.parser;

import calc.lexer.TokenType;

public class Precedence {
    public static final int LOWEST = 0;
    public static final int ADDITIVE = 10;
    public static final int MULTIPLICATIVE = 20;
    public static final int POWER = 30;

    public static int of(TokenType type) {
        switch (type) {
            case PLUS:
            case MINUS:
                return ADDITIVE;
            case STAR:
            case SLASH:
            case PERCENT:
                return MULTIPLICATIVE;
            case CARET:
                return POWER;
            default:
                return LOWEST;
        }
    }

    public static boolean rightAssociative(TokenType type) {
        return type == TokenType.CARET;
    }

    /** The minimum precedence for the right-hand side of the given operator. */
    public static int nextFor(TokenType type) {
        return rightAssociative(type) ? of(type) - 1 : of(type);
    }
}
